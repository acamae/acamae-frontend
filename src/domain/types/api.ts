import { UserRole } from '@domain/constants/user';

export interface UserResponse {
  id: string;
  email: string;
  username: string;
  role?: UserRole;
  createdAt: string;
  updatedAt: string;
}
