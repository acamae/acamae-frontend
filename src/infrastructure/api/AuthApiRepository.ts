import { AxiosError, AxiosResponse } from 'axios';

import {
  ApiPromise,
  ApiResponse,
  ForgotPasswordPayload,
  LoginPayload,
  RegisterPayload,
  ResetPasswordPayload,
} from '@/domain/types/apiSchema';
import { USER_ROLES } from '@domain/constants/user';
import { User } from '@domain/entities/User';
import { AuthRepository } from '@domain/repositories/AuthRepository';
import { UserResponse } from '@domain/types/api';
import {
  API_ROUTES,
  getUserByIdUrl,
  getUpdateUserByIdUrl,
  getDeleteUserByIdUrl,
  getAuthResetPasswordUrl,
} from '@shared/constants/apiRoutes';
import api from '@shared/services/axiosService';

function mapUserResponse(data: UserResponse): User {
  return {
    id: data.id,
    email: data.email,
    username: data.username,
    role: data.role === USER_ROLES.ADMIN ? USER_ROLES.ADMIN : USER_ROLES.USER,
    createdAt: new Date(data.createdAt),
    updatedAt: new Date(data.updatedAt),
  };
}

/**
 * Map the response from the AxiosResponse type to the ApiResponseSuccess type
 * @param response AxiosResponse
 * @description 2xx response with data
 * @returns ApiResponseSuccess
 */
function mapApiAxiosResponseError<T>({
  response,
}: {
  response: AxiosResponse<T>; // data, status, statusText, headers, config, request?
}): ApiResponse<T> {
  return {
    data: response.data,
    status: response.status,
    success: false,
  };
}

/**
 * Map the error from the AxiosError type to the ApiError type
 * @param error AxiosError
 * @description 4xx or 5xx response with error
 * @returns ApiError
 */
function mapApiAxiosRequestError<T>({
  error,
}: {
  error: AxiosError<T>; // response, request, status, cause, code, message, name, stack, toJSON
}): ApiResponse<T> {
  return {
    success: false,
    status: error.status ?? 500,
    message: error.message ?? 'A request error occurred',
    code: error.code ?? 'REQUEST_ERROR',
  };
}

/**
 * Map the unknown error from the AxiosError type to the ApiError type
 * @param error AxiosError
 * @description unknown error
 * @returns ApiError
 */
function mapApiAxiosUnknownError<T>(): ApiResponse<T> {
  return {
    success: false,
    status: 500,
    message: 'An unknown error occurred',
    code: 'UNKNOWN_ERROR',
  };
}

/**
 * Map the response from the AxiosResponse type to the ApiResponseSuccess type
 * @param response AxiosResponse
 * @description 2xx response with data
 * @returns ApiResponseSuccess
 */
function handleApiSuccess<T>({
  response,
}: {
  response: AxiosResponse<T>; // data, status, statusText, headers, config, request?
}): ApiResponse<T> {
  return {
    data: response.data,
    status: response.status,
    success: response.status >= 200 && response.status < 300,
  };
}

/**
 * Map the error from the AxiosError type to the ApiError type
 * @param error AxiosError
 * @description 4xx or 5xx response with error
 * @returns ApiError
 */
function handleApiError<T>(error: unknown): ApiResponse<T> {
  if (error instanceof AxiosError && error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    return mapApiAxiosResponseError({ response: error.response });
  } else if (error instanceof AxiosError && error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    return mapApiAxiosRequestError({ error });
  }

  // Something happened in setting up the request that triggered an Error
  return mapApiAxiosUnknownError();
}

export class AuthApiRepository implements AuthRepository {
  async login(payload: LoginPayload): ApiPromise<User> {
    try {
      const response = await api.post(API_ROUTES.AUTH.LOGIN, {
        ...payload,
      });
      return handleApiSuccess({
        response: {
          ...response,
          data: { ...mapUserResponse(response.data) },
        },
      });
    } catch (error: unknown) {
      return handleApiError(error);
    }
  }

  async register(payload: RegisterPayload): ApiPromise<User> {
    try {
      const response = await api.post(API_ROUTES.AUTH.REGISTER, payload);
      return handleApiSuccess({
        response: {
          ...response,
          data: { ...mapUserResponse(response.data) },
        },
      });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async forgotPassword(payload: ForgotPasswordPayload): ApiPromise<void> {
    try {
      const response = await api.post(API_ROUTES.AUTH.FORGOT_PASSWORD, payload);
      return handleApiSuccess({ response });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async resetPassword(payload: ResetPasswordPayload): ApiPromise<void> {
    try {
      const response = await api.post(getAuthResetPasswordUrl(payload.token), {
        newPassword: payload.newPassword,
      });
      return handleApiSuccess({ response });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async logout(): ApiPromise<void> {
    try {
      const response = await api.post(API_ROUTES.AUTH.LOGOUT);
      return handleApiSuccess({ response });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async getCurrentUser(): ApiPromise<User> {
    try {
      const response = await api.get(API_ROUTES.AUTH.ME);
      return handleApiSuccess({
        response: {
          ...response,
          data: { ...mapUserResponse(response.data) },
        },
      });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async findAll(): ApiPromise<User[]> {
    try {
      const response = await api.get(API_ROUTES.USERS.GET_ALL);
      return handleApiSuccess({
        response: {
          ...response,
          data: response.data.map(mapUserResponse),
        },
      });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async findById(id: string): ApiPromise<User | null> {
    try {
      const response = await api.get(getUserByIdUrl(id));
      return handleApiSuccess({
        response: {
          ...response,
          data: { ...mapUserResponse(response.data) },
        },
      });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async save(entity: User): ApiPromise<User> {
    try {
      const response = await api.put(getUpdateUserByIdUrl(entity.id), {
        email: entity.email,
        username: entity.username,
        role: entity.role,
      });
      return handleApiSuccess({
        response: {
          ...response,
          data: { ...mapUserResponse(response.data) },
        },
      });
    } catch (error) {
      return handleApiError(error);
    }
  }

  async delete(id: string): ApiPromise<void> {
    try {
      const response = await api.delete(getDeleteUserByIdUrl(id));
      return handleApiSuccess({ response });
    } catch (error) {
      return handleApiError(error);
    }
  }
}
