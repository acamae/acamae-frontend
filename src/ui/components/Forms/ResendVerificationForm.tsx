import React, { useState } from 'react';
import { Form, Button, Alert } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';

import { API_ROUTES } from '@shared/constants/apiRoutes';
import api from '@shared/services/axiosService';

const ResendVerificationForm: React.FC = () => {
  const { t } = useTranslation();
  const [identifier, setIdentifier] = useState('');
  const [status, setStatus] = useState<'idle' | 'success' | 'error' | 'loading'>('idle');
  const [message, setMessage] = useState('');

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    setStatus('loading');
    setMessage('');

    try {
      await api.post(API_ROUTES.AUTH.RESEND_VERIFICATION, { identifier });
      setStatus('success');
      setMessage(t('verification.resend.success'));
    } catch (err: unknown) {
      setStatus('error');
      let finalErrorMessage = t('errors.unknown');
      if (err instanceof Error && err.message) {
        if (err.message === 'USER_NOT_FOUND') {
          finalErrorMessage = t('errors.email.not_found');
        } else if (err.message === 'ALREADY_VERIFIED') {
          finalErrorMessage = t('email.verification_repeat');
        }
      }
      setMessage(finalErrorMessage);
    }
  };

  return (
    <>
      {status !== 'idle' && (
        <Alert
          variant={status === 'success' ? 'success' : 'danger'}
          data-testid={status === 'success' ? 'alert-success' : 'alert-error'}>
          {message}
        </Alert>
      )}

      <Form onSubmit={handleSubmit}>
        <Form.Group className="mb-3" controlId="identifier">
          <Form.Label>{t('verification.resend.label')}</Form.Label>
          <Form.Control
            type="text"
            value={identifier}
            onChange={e => setIdentifier(e.target.value.trim())}
            required
            autoFocus
            data-testid="input-email-username"
          />
        </Form.Group>

        <div className="d-grid">
          <Button
            variant="primary"
            type="submit"
            disabled={status === 'loading'}
            data-testid="btn-resend-verification">
            {status === 'loading'
              ? t('verification.resend.loading')
              : t('verification.resend.button')}
          </Button>
        </div>
      </Form>
    </>
  );
};

export default ResendVerificationForm;
