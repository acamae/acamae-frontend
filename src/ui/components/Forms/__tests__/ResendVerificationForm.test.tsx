import { screen, fireEvent, waitFor, render } from '@testing-library/react';

import api from '@shared/services/axiosService';

import ResendVerificationForm from '../ResendVerificationForm';

jest.mock('@shared/services/axiosService', () => ({
  post: jest.fn(),
}));

describe('ResendVerificationForm', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders input and button', () => {
    render(<ResendVerificationForm />);
    expect(screen.getByTestId('input-email-username')).toBeInTheDocument();
    expect(screen.getByTestId('btn-resend-verification')).toBeInTheDocument();
  });

  it('shows success message on successful resend', async () => {
    (api.post as jest.Mock).mockResolvedValueOnce({});
    render(<ResendVerificationForm />);
    fireEvent.change(screen.getByTestId('input-email-username'), {
      target: { value: 'test@example.com' },
    });
    fireEvent.click(screen.getByTestId('btn-resend-verification'));
    await waitFor(() => {
      expect(screen.getByTestId('alert-success')).toBeInTheDocument();
    });
  });

  it('shows error message on user not found', async () => {
    (api.post as jest.Mock).mockRejectedValueOnce(new Error('USER_NOT_FOUND'));
    render(<ResendVerificationForm />);
    fireEvent.change(screen.getByTestId('input-email-username'), {
      target: { value: 'notfound@example.com' },
    });
    fireEvent.click(screen.getByTestId('btn-resend-verification'));
    await waitFor(() => {
      expect(screen.getByTestId('alert-error')).toBeInTheDocument();
    });
  });

  it('shows error message on already verified', async () => {
    (api.post as jest.Mock).mockRejectedValueOnce(new Error('ALREADY_VERIFIED'));
    render(<ResendVerificationForm />);
    fireEvent.change(screen.getByTestId('input-email-username'), {
      target: { value: 'already@example.com' },
    });
    fireEvent.click(screen.getByTestId('btn-resend-verification'));
    await waitFor(() => {
      expect(screen.getByTestId('alert-error')).toBeInTheDocument();
    });
  });

  it('shows unknown error for other errors', async () => {
    (api.post as jest.Mock).mockRejectedValueOnce(new Error('SOME_OTHER_ERROR'));
    render(<ResendVerificationForm />);
    fireEvent.change(screen.getByTestId('input-email-username'), {
      target: { value: 'fail@example.com' },
    });
    fireEvent.click(screen.getByTestId('btn-resend-verification'));
    await waitFor(() => {
      expect(screen.getByTestId('alert-error')).toBeInTheDocument();
    });
  });

  it('snapshot render', () => {
    const { asFragment } = render(<ResendVerificationForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
