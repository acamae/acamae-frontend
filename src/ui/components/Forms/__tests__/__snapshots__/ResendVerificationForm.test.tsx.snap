// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResendVerificationForm snapshot render 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="identifier"
      >
        Correo electrónico o nombre de usuario
      </label>
      <input
        class="form-control"
        data-testid="input-email-username"
        id="identifier"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="d-grid"
    >
      <button
        class="btn btn-primary"
        data-testid="btn-resend-verification"
        type="submit"
      >
        Reenviar verificación
      </button>
    </div>
  </form>
</DocumentFragment>
`;
